pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go

```diff
@@ -72,10 +80,22 @@ func NewNamespaceController(kubeClient kube.Client, caBundleWatcher *keycertbund
 			// skip special kubernetes system namespaces
 			return false
 		}
+		if features.EnableEnhancedResourceScoping && !c.DiscoveryNamespacesFilter.Filter(o) {
+			// This is a change to a configmap we don't watch, ignore it
+			return false
+		}
 		return true
 	}))
 	c.namespacesInformer.AddEventHandler(controllers.FilteredObjectSpecHandler(c.queue.AddObject, func(o controllers.Object) bool {
-		return !inject.IgnoredNamespaces.Contains(o.GetName())
+		if inject.IgnoredNamespaces.Contains(o.GetName()) {
+			// skip special kubernetes system namespaces
+			return false
+		}
+		if features.EnableEnhancedResourceScoping && !c.DiscoveryNamespacesFilter.FilterNamespace(o.GetName()) {
```

Comment:
One concern...

consider a namespace is added:
* We have 2 informers for namespace, namespacecontroller and DiscoveryFilter
* Events are triggered in random order to these
* namespacecontroller gets "Add ns foo" event. It checks the discovery filter, and it is not present so it ignores it
* DiscoveryFilter gets "Add ns foo" event. It should be included, so we add it

Now we have failed to configure `foo`?

Reply:
Good point! Instead of utilizing the `DiscoveryNamespacesFilter.FilterNamespace`  which needs to watch the namespace creation and have the sequence issue, suppose we can just utilize the `DiscoveryNamespaceFilter.discoverySelectors`   to filter the new ns here?

Reply:
>Events are triggered in random order to these

Events is in determined order, because actually the two informers are sharing a underlying store

Reply:
>namespacecontroller gets "Add ns foo" event. It checks the discovery filter, and it is not present so it ignores it

IN order to prevent this, we should make sure the discover filter cb get executed before this one here

Reply:
simply we can merge both cb handler into one, currently two queues can be run in random order

Reply:
Seems simpler to do @irisdingbj's suggestion, then we are not dependent on tricky race conditions

Reply:
Done. @irisdingbj could you please have a relook?